pkg/inventory/db.go

```diff
@@ -138,17 +163,94 @@ func (db *DB) Run(ctx context.Context) error {
 	// cases.
 	gwInterfaces := getDefaultGwInterfaces()
 
+	// Run initial full discovery
+	if err := db.discoverDevices(true, gwInterfaces); err != nil {
+		klog.Errorf("Initial full device discovery failed: %v", err)
+	}
+
+	fullDiscoveryTicker := time.NewTicker(db.fullDiscoveryInterval)
+	defer fullDiscoveryTicker.Stop()
+
+	simpleDiscoveryTicker := time.NewTicker(db.simpleDiscoveryInterval)
+	defer simpleDiscoveryTicker.Stop()
+
+	nlChannel := make(chan netlink.LinkUpdate)
+	doneCh := make(chan struct{})
+	defer close(doneCh)
+	if err := netlink.LinkSubscribe(nlChannel, doneCh); err != nil {
+		klog.Error(err, "error subscribing to netlink interfaces, only syncing periodically", "interval", db.simpleDiscoveryInterval.String())
+	}
+
 	for {
 		err := db.rateLimiter.Wait(ctx)
 		if err != nil {
 			klog.Error(err, "unexpected rate limited error trying to get system interfaces")
 		}
 
-		devices := []resourceapi.Device{}
+		select {
+		case <-fullDiscoveryTicker.C:
+			if err := db.discoverDevices(true, gwInterfaces); err != nil {
+				klog.Errorf("full device discovery failed: %v", err)
+			}
+		case <-simpleDiscoveryTicker.C:
+			if err := db.discoverDevices(false, gwInterfaces); err != nil {
+				klog.Errorf("simple device discovery failed: %v", err)
+			}
+		case <-nlChannel:
+			// drain the channel so we only sync once
+			for len(nlChannel) > 0 {
+				<-nlChannel
+			}
+			if err := db.discoverDevices(false, gwInterfaces); err != nil {
+				klog.Errorf("simple device discovery failed: %v", err)
+			}
+		case <-ctx.Done():
+			return ctx.Err()
+		}
+	}
+}
+
+func (db *DB) discoverDevices(full bool, gwInterfaces sets.Set[string]) error {
+	var netNamespaces []netns.NsHandle
+	var err error
+
+	// Always scan the host network namespace
+	hostNs, err := netns.Get()
+	if err != nil {
+		return fmt.Errorf("failed to get host network namespace: %w", err)
+	}
+	netNamespaces = append(netNamespaces, hostNs)
+
+	if full {
+		// In case of full device discovery, also scan the pod's network namespaces.
+		for pod, nsPath := range db.listPodNetNamespaces() {
+			if nsPath == "" {
+				continue
+			}
+			ns, err := netns.GetFromPath(nsPath)
+			if err != nil {
+				klog.Errorf("Failed to get netns handle for pod %q having network namespace %q: %v", pod, nsPath, err)
+				continue
+			}
+			netNamespaces = append(netNamespaces, ns)
+		}
+	}
+
+	newDeviceStore := make(map[string]*resourceapi.Device)
+
+	for _, ns := range netNamespaces {
+		nlHandle, err := netlink.NewHandleAt(ns)
+		if err != nil {
+			klog.Errorf("failed to create netlink handle for netns %v: %v", ns, err)
+			continue
+		}
+
```

Comment:
I do not like much this idea of getting into the network namespaces, it is very invasive, why do we need to do this?